signer/signer.go

```diff
@@ -0,0 +1,249 @@
+	if err := validateSignRequest(req); err != nil {
+		return nil, err
+	}
+
+	sig, err := s.internalEnvelope.signPayload(req)
+	if err != nil {
+		return nil, err
+	}
+
+	s.rawSignatureEnvelope = sig
+	return sig, err
+}
+
+// GetSignerInfo returns information about the Signature envelope
+func (s SignatureEnvelope) GetSignerInfo() (SignerInfo, error) {
+	if len(s.rawSignatureEnvelope) == 0 {
+		return SignerInfo{}, SignatureNotFoundError{}
+	}
+
+	signInfo, err := s.internalEnvelope.getSignerInfo()
+	if err != nil {
+		return signInfo, err
+	}
+	if err := validateSignerInfo(signInfo); err != nil {
+		return signInfo, err
+	}
+	return signInfo, nil
+}
+
+// validateSignerInfo performs basic set of validations on SignerInfo struct.
+func validateSignerInfo(info SignerInfo) error {
+	if len(info.Payload) == 0 {
+		return MalformedSignatureError{msg: "payload not present"}
+	}
+
+	// TODO: perform PayloadContentType value validations
+	if len(info.PayloadContentType) == 0 {
+		return MalformedSignatureError{msg: "signature content type not present or is empty"}
+	}
+
+	signTime := info.SignedAttributes.SigningTime
+	if signTime.IsZero() {
+		return MalformedSignatureError{msg: "singing-time not present"}
+	}
+
+	expTime := info.SignedAttributes.Expiry
+	if !expTime.IsZero() {
+		if expTime.Before(signTime) || expTime.Equal(signTime) {
+			return MalformedSignatureError{msg: "expiry cannot be equal or before the signing time"}
+		}
+	}
+
+	if len(info.CertificateChain) == 0 {
+		return MalformedSignatureError{msg: "certificate-chain not present or is empty"}
+	}
+
+	if len(info.Signature) == 0 {
+		return MalformedSignatureError{msg: "signature not present or is empty"}
+	}
+
+	return nil
+}
+
+// validateSignRequest performs basic set of validations on SignRequest struct.
+func validateSignRequest(req SignRequest) error {
+	if len(req.Payload) == 0 {
+		return MalformedSignRequestError{msg: "payload not present"}
+	}
+
+	// TODO: perform PayloadContentType value validations
+	if len(req.PayloadContentType) == 0 {
+		return MalformedSignRequestError{msg: "signature content type not present or is empty"}
+	}
+
+	if len(req.CertificateChain) == 0 {
+		return MalformedSignRequestError{msg: "certificate-chain not present or is empty"}
+	}
+
+	signTime := req.SigningTime
+	if signTime.IsZero() {
+		return MalformedSignRequestError{msg: "singing-time not present"}
+	}
+
+	expTime := req.Expiry
+	if !expTime.IsZero() {
+		if expTime.Before(signTime) || expTime.Equal(signTime) {
+			return MalformedSignRequestError{msg: "expiry cannot be equal or before the signing time"}
+		}
+	}
+
+	if req.SignatureProvider == nil {
+		return MalformedSignRequestError{msg: "SignatureProvider is nil"}
+	}
+	return nil
+}
+
+// NewSignatureEnvelopeFromBytes is used for signature verification workflow
+func NewSignatureEnvelopeFromBytes(envelopeBytes []byte, envelopeMediaType signatureMediaType) (SignatureEnvelope, error) {
```

Comment:
According to the context, `SignatureEnvelope` should be an interface.

Reply:
It cannot be an interface as we are storing state(the signature).